File: tp-5/Queue.hs
module Queue where

data Queue a = Q [a]
  deriving Show

-- Una Queue es un tipo abstracto de datos de naturaleza FIFO (first in, first out). Esto significa
-- que los elementos salen en el orden con el que entraron, es decir, el que se agrega primero es el
-- primero en salir (como la cola de un banco).

-- Crea una cola vacía.
-- O(1) ya que la operacion es fija, no crece segun sus parametros (en este caso no tiene)
emptyQ :: Queue a
emptyQ = Q []

-- Dada una cola indica si la cola está vacía.
-- O(1) ya que la operacion es fija, no crece segun sus parametros
-- en este caso es un pattern matching de igualdad
isEmptyQ :: Queue a -> Bool
isEmptyQ (Q []) = True
isEmptyQ _      = False

-- Dados un elemento y una cola, agrega ese elemento a la cola.
-- O(n) donde n es la cantidad de elementos de la lista (l)
-- es de costo lineal ya que para agregar un elemento al final de la lista
-- se debe atravesar toda la lista
enqueue :: a -> Queue a -> Queue a
enqueue a (Q l) = Q (l ++ [a])

-- Dada una cola devuelve el primer elemento de la cola.
-- O(1) ya que head es una operacion de costo constante
firstQ :: Queue a -> a
firstQ (Q l) = head l

-- Dada una cola la devuelve sin su primer elemento.
-- O(1) ya que tail es una operacion de costo constante
dequeue :: Queue a -> Queue a
dequeue (Q l) = Q (tail l)

File: tp-11/Queue.cpp
#include "Queue.h"
#include "iostream"

/* Crea una lista vacía. */
/* Costo: O(1). */
Queue emptyQ() {
  Queue q = new QueueSt;
  q->cantidad = 0;
  q->primero = NULL;
  q->ultimo = NULL;
  return q;
}

/* Indica si la lista está vacía. */
/* Costo: O(1). */
bool isEmptyQ(Queue q) { return q->cantidad == 0; }

/* Devuelve el primer elemento. */
/* Costo: O(1). */
int firstQ(Queue q) { return q->primero->elem; }

/* Agrega un elemento al final de la cola. */
/* Costo: O(1). */
void Enqueue(int x, Queue q) {
  NodoQ *n = new NodoQ;
  n->elem = x;
  n->siguiente = NULL;

  q->ultimo->siguiente = n;
}

/* Quita el primer elemento de la cola. */
/* Costo: O(1). */
void Dequeue(Queue q) {
  delete q->primero;
  q->primero = q->primero->siguiente;
}

/* Devuelve la cantidad de elementos de la cola. */
/* Costo: O(1). */
int lengthQ(Queue q) { return q->cantidad; }

/* Anexa q2 al final de q1, liberando la memoria inservible de q2 en el proceso.
 */
/* Nota: Si bien se libera memoria de q2, no necesariamente la de sus nodos. */
/* Costo: O(1). */
void MergeQ(Queue q1, Queue q2) {
  q1->ultimo = q2->primero;
  delete q2;
}

/* Libera la memoria ocupada por la lista. */
/* Costo: O(n). */
void DestroyQ(Queue q) {
  while (!isEmptyQ(q)) {
    delete q->primero;
    Dequeue(q);
  }
  delete q;
}
